### PR TITLE
Fix undefined temporary environment restoring issue

### LINKED ISSUE #1448: Unset environment variable after temporary assignment
After temporary assignment to an environment variable, that variable should be unset after the function exit, _provided_ it was unset before. Consider this (using the new `tmp` syntax):

```
⬥ date; has-env TZ; { tmp E:TZ = UTC; date }; date; has-env TZ; put $E:TZ
Sat Dec 11 10:07:40 CET 2021
⮕ $false
Sat Dec 11 09:07:40 UTC 2021
Sat Dec 11 09:07:40 UTC 2021
⮕ $true
⮕ ''
```

As can be seen from the various `date` outputs, it does make a difference whether an environment variable is unset or merely empty.

### LINKED ISSUE #1536: Setting temporary environment variable will leave a blank environment behind
```sh
~> env | grep TEST
Exception: grep exited with 1
[tty 120], line 1: env | grep TEST
~> { tmp E:TEST = 1; put $E:TEST }
▶ 1
~> env | grep TEST
TEST=
```

### PR DESCRIPTION
**Pull request by tw4452852**
This is the 2nd attempt to fix this issue. The previous one is #1537 

Close #1448 

Signed-off-by: Tw <tw19881113@gmail.com>

### CODE REVIEW

File: pkg/eval/builtin_special_test.go [hunk 1]
```diff
@@ -147,6 +147,11 @@ func TestTmp(t *testing.T) {
 
 		That("var x; tmp x = y").DoesNotCompile(),
 		That("{ tmp x = y }").DoesNotCompile(),
+
+		That("{ tmp E:X = y; put $E:X }; has-env X").
+			Puts("y", false),
```
Comment (krader1961) [inline pkg/eval/builtin_special_test.go:152]: AFAICT there is no test that expanding an undefined env var returns an empty string. This is a good time to add such a test and this test seems like a natural place to do it. Simple add `; put $E:X` to the end of the test and `""` as the third expected value.

Comment (krader1961) [inline pkg/eval/builtin_special_test.go:152]: Also, there doesn't appear to be an equivalent test of `set` so this is a good time to add one. For example:

```
That("{ set E:X = y; put $E:X }; has-env X; put $E:X").
    Puts("y", true, "y"),
```

File: pkg/eval/vars/env.go [hunk 2]
```diff
@@ -7,23 +7,32 @@ import (
...
-func (ev envVariable) Get() any {
+func (ev EnvVariable) Get() any {
 	return os.Getenv(ev.name)
 }
 
+func (ev EnvVariable) Unset() error {
+	return os.Unsetenv(ev.name)
+}
+
+func (ev EnvVariable) IsExist() bool {
```
Comment (krader1961) [inline pkg/eval/vars/env.go:30]: This method name is awkward. Maybe change it to simply `Exists` or `IsSet`.

**Comment by krader1961** [comment]
LGTM from a behavior perspective, and modulo my comments about improving testing of this behavior and the naming of one new method.

File: pkg/eval/builtin_special_test.go (hunk 1 above)
Comment (tw4452852) [reply pkg/eval/builtin_special_test.go]: I think `TestSet` is a better place, so I add this case there.

**Review by xiaq — APPROVED** [review]

**Pull request merged**